### ISSUE TITLE
Restructure of Mapping classes (into `mars-sim-mapdata`)

### ISSUE THREAD

**Issue by bdowrick** [mars-sim/mars-sim#947]
**Is your feature request related to a problem? Please describe.**

Currently in the `mars-sim` project, a number of map features are within the `mars-sim-core` (or worse, the `mars-sim-ui`) module and allow for usage of such things as `Coordinates`, `LocalPosition`, `LocalAreaUtil`, `OrbitInfo` for location and information.

Some features of `mars-sim-mapdata` are duplicated over into `mars-sim-ui` as well, such as `Map` vs `MapData` which are near identical in usage, with minor changes between them.

**Describe the solution you desire to see**

A complete restructure of _all_ geo/positional related tools and classes into `mars-sim-mapdata` and create util/tool classes for `mars-sim-core` to extend and populate the world with data.

_Side note:_
There are some shared classes that would also need to be moved over into an additional module `mars-sim-tools`

**Describe the rationale**

While this is a considerable change, and may require a fair bit of work to actually implement, I believe it's the best way to move forward with `mars-sim` as it allow for reduced code duplication, lets the `mars-sim-core` focus solely on the model of the simulation, and `mars-sim-ui` focus on UI.

**Screenshots**

![image](https://github.com/mars-sim/mars-sim/assets/16632004/ab9fb032-e5ae-46c2-a68b-d3a96fc4fac4)

**Comment by bdowrick** [comment]
@mokun @bevans2000 

Working through #935 has lead me into thinking this may be a better way of doing things, as more than just `MineralMapLayer` would benefit. Allowing me to improve performance of all layers (and anything location related) a lot easier.

(Sorry for delay as well, I've been sick past few days)

**Comment by bevans2000** [comment]
Totally agree. I have been thinking something similar for a while..
The logging classes and directory management should go into this tools package.
I think the change isn't too bad, obviously it needs to be done on a branch

**Comment by mokun** [comment]
Hope you feel better with good reat.

Yes let's create mars-sim-tool submodule and  separate concerns better and make things cleaner.

**Comment by mokun** [comment]
Good that `mars-sim-tools` is created.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/21f41b53-1234-4806-aa75-796ef7dc6b06)

So far, there are only two class files in it.

Any other class going there as well ?

**Comment by bdowrick** [comment]
I plan on moving OpenCL to it, alongside any other classes that may be useful to other modules.

I kind of like the idea of making `mars-sim-core` only responsible for the simulation model, and any techincal details, like saving/loading, preference management to be in `mars-sim-tools`.

What are your thoughts?

**Comment by mokun** [comment]
Those are good idea. 

A side question here. Any other areas of computation you see may adopt JOCL ?

**Comment by bdowrick** [comment]
Yes I have :) My focus has only been on the mapping related sections to find better ways of managing the data

My biggest goal is to get all map layers rendering at the same time in 16ms or less. It's taking a bit of restructuring and time to read through the classes, to do but there's a number of applications for JOCL to help do some heavy lifting

**Comment by bevans2000** [comment]
Yes I agree that core should be about running the simulation.  Saving & Loading could be tricky to separate but things like the default directory logic we can move. 
We probably need to tackle the package structure of classes that are moved as well.

**Comment by mokun** [comment]
> there's a number of applications for JOCL to help do some heavy lifting

I can see that [AirComposition](https://github.com/mars-sim/mars-sim/blob/master/mars-sim-core/src/main/java/org/mars_sim/msp/core/air/AirComposition.java) would be benefit from JOCL.

In the past, it used to be setup with double array for handling all those thermodynamic (using ideal gas law) calculation. 

However, with array computation, it was harder to debug. So Barry convert all of them to iterating them through Set, List and Map. 

Now it's probably good to go back and look to the old code in history and convert that into using JOCL. 

If GPU is found on a player machine, it will switch to using GPU to do the heavy lifting.

**Comment by bdowrick** [comment]
I was looking into that the other day. I had to research a bit onto the actual math behind it, but I believe it'll be quite easy to implement into GPU. I haven't given it a go yet but I can an issue and discuss it more when I get started on it?

There's also some simple performance tricks that can be used to speed it up a little bit more for devices that can't utilize JOCL, and before the rollout of the performance overhaul
